File: ios/GradeThread/Capture/PhotoProfile.swift
import Foundation
import Observation

/// Client mirror of the server-authoritative photo profiles
/// (services/edge-functions/src/lib/photo-profiles.ts). A profile maps an
/// `item_category` to the ordered photo "roles" a seller should capture, with
/// category-specific labels, hints, and required flags — so a watch is asked
/// for a dial/caseback/serial instead of a garment's front/back/tag/detail.
///
/// The full table is fetched from `GET /api/flipdesk/photo-profiles` via
/// ``PhotoProfileStore`` and cached; the small bundled fallback below keeps the
/// UI usable offline / before the first fetch. The server table always wins
/// once it loads, so the fallback only needs to be "good enough".

public struct PhotoRole: Decodable, Hashable {
    /// Physical `item_photos.photo_type` value this slot writes.
    public let type: String
    /// US-2468 (migration 00587): the `item_photos.photo_role` qualifier this
    /// slot writes, or nil for a slot that takes none. Slot identity is
    /// (type, role) — that is what lets a suit hold three separate `tag` slots
    /// instead of needing a `tag_2` and a `tag_3` in the enum.
    public let role: String?
    /// Category-specific display label, e.g. "Dial / Face", "Front of card".
    public let label: String
    /// One-line capture guidance.
    public let hint: String
    /// Must be present before the item can advance to "photographed".
    public let required: Bool
    /// lucide icon name (web); iOS maps the slot to an SF Symbol via PhotoSlotType.
    public let icon: String
}

public struct PhotoProfile: Decodable, Hashable, Identifiable {
    /// `item_category` value this profile applies to.
    public let category: String
    /// Human category name for headers.
    public let label: String
    /// Ordered capture roles — required ones first by convention.
    public let roles: [PhotoRole]

    public var id: String { category }

    /// Whether this category offers defect close-ups.
    public var allowsDefects: Bool {
        roles.contains { $0.type == "defect" }
    }

    /// The role definition for a given server photo type, if the profile has one.
    public func role(forServerType serverType: String) -> PhotoRole? {
        roles.first { $0.type == serverType }
    }

    /// US-2468: the role definition for a stored (type, role) PAIR. Prefer this
    /// over ``role(forServerType:)`` whenever the caller has both halves — a
    /// suit profile has three `tag` slots and the type alone picks the wrong
    /// one two times out of three.
    public func role(forServerType serverType: String, photoRole: String?) -> PhotoRole? {
        roles.first { $0.type == serverType && $0.role == photoRole }
            ?? roles.first { $0.type == serverType && $0.role == nil }
    }

    /// Stable identity for a slot: "type" or "type:role".
    public static func slotKey(_ type: String, _ role: String?) -> String {
        guard let role, !role.isEmpty else { return type }
        return "\(type):\(role)"
    }

    // MARK: - Bundled fallback

    /// US-2470: the fallback now carries ROLES, not just the four base slots.
    ///
    /// It used to hold five type-only entries, which was fine while the capture
    /// strip's "Add more" menu came from a hard-coded `PhotoSlotType.extras`
    /// list. That list is gone — the menu is built from whatever profile is
    /// resolved — so a bare fallback would leave a seller who opens the camera
    /// offline with nothing behind "Add more" except Defect. These mirror the
    /// server's clothing roles closely enough to shoot a garment properly, and
    /// the server table replaces them the moment it loads.
    public static let clothingFallback = PhotoProfile(
        category: "clothing",
        label: "Clothing",
        roles: [
            PhotoRole(type: "front", role: nil, label: "Front", hint: "Lay flat, full front in frame", required: true, icon: "shirt"),
            PhotoRole(type: "back", role: nil, label: "Back", hint: "Same crop as the front shot", required: true, icon: "shirt"),
            PhotoRole(type: "tag", role: "brand", label: "Brand label", hint: "The maker's logo or wordmark", required: false, icon: "tag"),
            PhotoRole(type: "tag", role: "size", label: "Size tag", hint: "The size itself, close enough to read without zooming", required: false, icon: "tag"),
            PhotoRole(type: "tag", role: "care", label: "Care & fabric", hint: "The care label with the fibre content", required: false, icon: "tag"),
            PhotoRole(type: "detail", role: "fabric", label: "Fabric close-up", hint: "Fill the frame with the weave or knit, in even light", required: false, icon: "search"),
            PhotoRole(type: "detail", role: "hardware", label: "Hardware", hint: "Zip pull, buttons, rivets or snaps", required: false, icon: "search"),
            PhotoRole(type: "detail", role: "print", label: "Print or graphic", hint: "The graphic straight on, close enough to show cracking", required: false, icon: "search"),
            PhotoRole(type: "measurement", role: nil, label: "Measurement card", hint: "Whole garment flat with the MeasureCard BESIDE it, shot top-down", required: false, icon: "ruler"),
            PhotoRole(type: "defect", role: nil, label: "Defect", hint: "Tight crop on any flaw — be honest", required: false, icon: "alert-triangle"),
            PhotoRole(type: "interior", role: nil, label: "Interior / Lining", hint: "Inside-out: lining, seams, interior tags", required: false, icon: "layers"),
            PhotoRole(type: "detail", role: "hem", label: "Hem & stitching", hint: "A hem or seam up close", required: false, icon: "search"),
            PhotoRole(type: "tag", role: "made_in", label: "Made in / union label", hint: "Origin, union or RN label", required: false, icon: "tag"),
            PhotoRole(type: "flatlay", role: nil, label: "Flat lay", hint: "Styled flat lay for the listing gallery", required: false, icon: "layout-grid"),
            PhotoRole(type: "on_hanger", role: nil, label: "On hanger", hint: "Hung straight on, showing how it drapes", required: false, icon: "shirt"),
            PhotoRole(type: "on_model", role: nil, label: "On model", hint: "Worn on a model or mannequin", required: false, icon: "user"),
        ]
    )

    public static let genericFallback = PhotoProfile(
        category: "other",
        label: "Other",
        roles: [
            PhotoRole(type: "front", role: nil, label: "Front", hint: "Main view in frame", required: true, icon: "image"),
            PhotoRole(type: "back", role: nil, label: "Back", hint: "Reverse view", required: true, icon: "image"),
            PhotoRole(type: "detail", role: nil, label: "Detail", hint: "Distinguishing detail or label", required: false, icon: "search"),
            PhotoRole(type: "defect", role: nil, label: "Defect", hint: "Tight crop on any flaw — be honest", required: false, icon: "alert-triangle"),
        ]
    )

    /// US-2812: mirrors SHOES in photo-profiles.ts verbatim.
    ///
    /// The bundled table covered `clothing` and `other` only, so a shoe
    /// captured while the profile fetch was unavailable — offline, first
    /// launch, a failed request — was offered Front/Back/Detail instead of
    /// Top/Heel/3/4 Angle/Sole/Size Stamp/Insole. The seller then has a set of
    /// photos that cannot show a sole, which is the one surface a shoe buyer
    /// looks at first.
    ///
    /// A DEGRADED-MODE path, not the normal one: the server table is the
    /// source of truth and is fetched per session. This only decides what
    /// happens when that fetch has not answered.
    public static let shoesFallback = PhotoProfile(
        category: "shoes",
        label: "Shoes",
        roles: [
            PhotoRole(type: "front", role: nil, label: "Top / Toe", hint: "Top-down or front; show both shoes if a pair", required: true, icon: "footprints"),
            PhotoRole(type: "back", role: nil, label: "Heel", hint: "Back of the heel, both shoes", required: true, icon: "footprints"),
            PhotoRole(type: "angle", role: nil, label: "3/4 Angle", hint: "Angled side view showing the silhouette", required: true, icon: "footprints"),
            PhotoRole(type: "sole", role: nil, label: "Sole", hint: "Outsole / tread — show wear honestly", required: true, icon: "footprints"),
            PhotoRole(type: "tag", role: nil, label: "Size Stamp", hint: "Tongue or insole size + brand stamp", required: false, icon: "tag"),
            PhotoRole(type: "interior", role: nil, label: "Insole", hint: "Inside the shoe — footbed condition", required: false, icon: "layers"),
            PhotoRole(type: "accessory", role: nil, label: "Box / Extras", hint: "Original box, spare laces, papers", required: false, icon: "package"),
            PhotoRole(type: "defect", role: nil, label: "Defect", hint: "Tight crop on any flaw — stain, snag, scuff, crack. Be honest.", required: false, icon: "alert-triangle"),
        ]
    )

    public static let bundledFallback: [String: PhotoProfile] = [
        "clothing": clothingFallback,
        "shoes": shoesFallback,
        "other": genericFallback,
    ]
}

/// Fetches + caches the photo-profile table. Static config → a single fetch
/// per session (server-cached too). Falls back to the bundled profiles when
/// the network is unavailable so callers can always resolve a profile.
///
/// US-2496: the table is NOT account-blind. `GET /api/flipdesk/photo-profiles`
/// answers per `workspaceOwnerId ?? userId` - a seller who cannot use the
/// authenticity add-on is served fewer roles - so a table fetched for one tenant
/// is the wrong answer for the next one. This store is created once in
/// ``GradeThreadApp`` and outlives both a sign-out and a workspace switch, so
/// the table carries the owner it was fetched FOR and is only ever handed back
/// to that owner. A stale table reads as "not loaded": callers get the bundled
/// fallback and the next ``loadIfNeeded()`` refetches.
///
/// The stamp is the mechanism deliberately, rather than a reset called from the
/// sign-out and workspace-switch paths. A stamp cannot be forgotten by a future
/// exit path, and this cache has three of them already (sign-out, workspace
/// switch, and the involuntary switch when a membership is revoked mid-session).
@MainActor
@Observable
public final class PhotoProfileStore {
    /// A loaded table plus the tenant it was loaded for.
    private struct Loaded {
        let owner: String
        let profiles: [String: PhotoProfile]
    }

    private var loaded: Loaded?
    private let ownerProvider: @MainActor () -> String?

    /// The server table when it belongs to the CURRENT tenant, else the bundled
    /// fallback. Checked on every read rather than only at load time, because
    /// the surfaces that display a profile (the item canvas, the photo manager)
    /// do not all call ``loadIfNeeded()`` first - a check that only ran on load
    /// would leave them showing the previous account's table until capture was
    /// opened.
    public var profiles: [String: PhotoProfile] {
        guard let owner = ownerProvider(), let loaded, loaded.owner == owner else {
            return PhotoProfile.bundledFallback
        }
        return loaded.profiles
    }

    /// - Parameter ownerProvider: active workspace owner, else the signed-in
    ///   user; nil when signed out. Injectable for previews and tests; nil
    ///   takes the live one below.
    ///
    /// The live provider is built in the BODY rather than written as a default
    /// argument, and it has to be. A default argument on a `public` declaration
    /// may only name `public` symbols, and `WorkspaceScope` is internal — which
    /// is what broke the iOS build on 2026-08-11: "enum 'WorkspaceScope' is
    /// internal and cannot be referenced from a default argument value", twice,
    /// and every Build + test run since died there.
    ///
    /// Moving it also drops four MemberImportVisibility warnings. This file
    /// imports neither Supabase nor Auth, and `.auth.currentUser` names members
    /// of both; that is only diagnosed in default-argument position, and it is
    /// an error rather than a warning under the Swift 6 language mode.
    public init(ownerProvider: (@MainActor () -> String?)? = nil) {
        self.ownerProvider = ownerProvider ?? {
            WorkspaceScope.activeOwnerId ?? SupabaseShared.client.auth.currentUser?.id.uuidString
        }
    }

    /// Resolves the profile for an `item_category`, falling back to clothing
    /// (the historical default) for null/unknown categories.
    public func profile(for category: String?) -> PhotoProfile {
        let table = profiles
        if let category, let match = table[category] {
            return match
        }
        return table["clothing"] ?? PhotoProfile.clothingFallback
    }

    /// US-2468: resolves the profile for an item, consulting the free-text
    /// GARMENT word ("blazer", "dress pants") as well as the item_category.
    ///
    /// Mirrors `getPhotoProfile(category, garment)` on the edge, including its
    /// fallback: when no usable item_category is given, the garment word is
    /// used to resolve a group and that group is mapped back to an
    /// item_category profile where one exists. That matters because
    /// `items_full` has no item_category column at all, so the free-text word
    /// is often the only category signal a caller actually has.
    public func profile(for category: String?, garment: String?) -> PhotoProfile {
        // One snapshot for the whole resolution: re-reading the tenant between
        // the lookups could mix two tables in one answer.
        let table = profiles
        if let category, category != "clothing", let match = table[category] {
            return match
        }
        let group = GarmentGroup.from(garment ?? category)
        if let key = group.itemCategoryProfileKey, let match = table[key] {
            return match
        }
        return table[group.clothingProfileKey]
            ?? table["clothing"]
            ?? PhotoProfile.clothingFallback
    }

    /// Loads the table from the edge once PER TENANT. Safe to call repeatedly.
    /// Signed out there is no tenant to load for, so this is a no-op and callers
    /// keep the bundled fallback.
    public func loadIfNeeded() async {
        guard let owner = ownerProvider() else { return }
        guard loaded?.owner != owner else { return }
        struct Wrapper: Decodable { let profiles: [String: PhotoProfile] }
        do {
            let resp: Wrapper = try await EdgeAPI.shared.getJSON(
                "/api/flipdesk/photo-profiles",
                cacheTTL: 3600
            )
            // Stamped with the owner resolved BEFORE the request. A switch that
            // landed mid-flight leaves this table stamped for the tenant it was
            // actually fetched for, so the new tenant reads it as stale and
            // refetches - rather than inheriting an answer meant for someone else.
            if !resp.profiles.isEmpty {
                loaded = Loaded(owner: owner, profiles: resp.profiles)
            }
        } catch {
            // Keep the bundled fallback — non-fatal; we retry next launch.
        }
    }
}

File: android/app/src/main/java/com/gradethread/app/capture/PhotoProfile.kt
package com.gradethread.app.capture

import com.gradethread.app.platform.net.EdgeApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * US-1326: per-category photo roles, server-authoritative
 * (services/edge-functions/src/lib/photo-profiles.ts; iOS PhotoProfile). A
 * watch gets watch-shaped slots instead of clothing defaults. Fetched from
 * `GET /api/flipdesk/photo-profiles` (1h TTL via EdgeApi's cache); the small
 * bundled fallback keeps capture working offline / on fetch failure.
 */
@Serializable
data class PhotoRole(
    /** Server photo_type string. */
    val type: String,
    val label: String,
    val hint: String,
    val required: Boolean,
    val icon: String,
    /**
     * US-2469 (migration 00587): the `item_photos.photo_role` qualifier this
     * slot writes, or null for a slot that takes none. Slot identity is
     * (type, role) — that is what lets a suit hold three separate `tag` slots
     * instead of needing a `tag_2` and a `tag_3` in the enum.
     *
     * Defaulted so a profile served by an older edge still decodes, and so the
     * bundled fallbacks below stay readable.
     */
    val role: String? = null,
)

@Serializable
data class PhotoProfile(
    val category: String,
    val label: String,
    /** Ordered — display order in the strip. */
    val roles: List<PhotoRole>,
) {

    /** Required roles resolved to capture slots (unknown types dropped). */
    val requiredSlots: List<PhotoSlotType>
        get() = roles.filter { it.required }.mapNotNull { PhotoSlotType.fromWire(it.type) }

    /**
     * Optional, NON-DEFECT roles resolved to slots. Defects keep their own
     * reveal-one-at-a-time mechanism in the capture flow (category-agnostic
     * by design, AC3) — they are excluded here.
     */
    val optionalSlots: List<PhotoSlotType>
        get() = roles
            .filter { !it.required && it.type != "defect" }
            .mapNotNull { PhotoSlotType.fromWire(it.type) }

    /** Whether this category offers defect close-ups at all. */
    val allowsDefects: Boolean
        get() = roles.any { it.type == "defect" }

    fun roleForServerType(serverType: String): PhotoRole? =
        roles.firstOrNull { it.type == serverType }

    /**
     * US-2469: the role definition for a stored (type, role) PAIR. Prefer this
     * over [roleForServerType] whenever the caller has both halves — a suit
     * profile has three `tag` slots and the type alone picks the wrong one two
     * times out of three.
     */
    fun roleFor(serverType: String, photoRole: String?): PhotoRole? =
        roles.firstOrNull { it.type == serverType && it.role == photoRole }
            ?: roles.firstOrNull { it.type == serverType && it.role == null }

    companion object {
        /** Stable identity for a slot: "type" or "type:role". */
        fun slotKey(type: String, role: String?): String =
            if (role.isNullOrEmpty()) type else "$type:$role"

        /** Splits a [slotKey] back into its halves. */
        fun parseSlotKey(slot: String): Pair<String, String?> {
            val i = slot.indexOf(':')
            return if (i == -1) slot to null else slot.substring(0, i) to slot.substring(i + 1)
        }

        /**
         * US-2498: the bundled copy of the SERVER's clothing profile, role for
         * role and in the same order — the mirror of
         * `ios/GradeThread/Capture/PhotoProfile.swift` and of the `clothing`
         * entry in `src/lib/photo-profiles.ts`.
         *
         * It was five unroled slots, which mattered once the capture strip
         * started reading from here: a seller who opened the camera offline was
         * offered front, back, one nameless tag and one nameless detail, and the
         * Add menu had nothing in it but defects. A fallback is the profile a
         * seller gets on a bad connection, so it has to be the real one.
         */
        val clothingFallback = PhotoProfile(
            category = "clothing",
            label = "Clothing",
            roles = listOf(
                PhotoRole("front", "Front", "Lay flat, full front in frame", required = true, icon = "shirt"),
                PhotoRole("back", "Back", "Same crop as the front shot", required = true, icon = "shirt"),
                PhotoRole("tag", "Brand label", "The maker's logo or wordmark", required = false, icon = "tag", role = "brand"),
                PhotoRole("tag", "Size tag", "The size itself, close enough to read without zooming", required = false, icon = "tag", role = "size"),
                PhotoRole("tag", "Care & fabric", "The care label with the fibre content", required = false, icon = "tag", role = "care"),
                PhotoRole("detail", "Fabric close-up", "Fill the frame with the weave or knit, in even light", required = false, icon = "search", role = "fabric"),
                PhotoRole("detail", "Hardware", "Zip pull, buttons, rivets or snaps", required = false, icon = "search", role = "hardware"),
                PhotoRole("detail", "Print or graphic", "The graphic straight on, close enough to show cracking", required = false, icon = "search", role = "print"),
                PhotoRole("measurement", "Measurement card", "Whole garment flat with the MeasureCard BESIDE it, shot top-down", required = false, icon = "ruler"),
                PhotoRole("defect", "Defect", "Tight crop on any flaw — be honest", required = false, icon = "alert-triangle"),
                PhotoRole("interior", "Interior / Lining", "Inside-out: lining, seams, interior tags", required = false, icon = "layers"),
                PhotoRole("detail", "Hem & stitching", "A hem or seam up close", required = false, icon = "search", role = "hem"),
                PhotoRole("tag", "Made in / union label", "Origin, union or RN label", required = false, icon = "tag", role = "made_in"),
                PhotoRole("flatlay", "Flat lay", "Styled flat lay for the listing gallery", required = false, icon = "layout-grid"),
                PhotoRole("on_hanger", "On hanger", "Hung straight on, showing how it drapes", required = false, icon = "shirt"),
                PhotoRole("on_model", "On model", "Worn on a model or mannequin", required = false, icon = "user"),
            ),
        )

        val genericFallback = PhotoProfile(
            category = "other",
            label = "Other",
            roles = listOf(
                PhotoRole("front", "Front", "Main view in frame", required = true, icon = "image"),
                PhotoRole("back", "Back", "Reverse view", required = true, icon = "image"),
                PhotoRole("detail", "Detail", "Distinguishing detail or label", required = false, icon = "search"),
                PhotoRole("defect", "Defect", "Tight crop on any flaw — be honest", required = false, icon = "alert-triangle"),
            ),
        )

        /**
         * US-2812: mirrors SHOES in photo-profiles.ts verbatim.
         *
         * The bundled table covered `clothing` and `other` only, so a shoe
         * captured while the profile fetch was unavailable was offered
         * Front/Back/Detail instead of Top/Heel/3/4 Angle/Sole/Size
         * Stamp/Insole — leaving the seller with photos that cannot show a
         * sole, the one surface a shoe buyer looks at first.
         *
         * A DEGRADED-MODE path. The server table is the source of truth and
         * is fetched per session; this only decides what happens before it
         * answers.
         */
        val shoesFallback = PhotoProfile(
            category = "shoes",
            label = "Shoes",
            roles = listOf(
                PhotoRole(
                    "front", "Top / Toe", "Top-down or front; show both shoes if a pair",
                    required = true, icon = "footprints",
                ),
                PhotoRole("back", "Heel", "Back of the heel, both shoes", required = true, icon = "footprints"),
                PhotoRole(
                    "angle", "3/4 Angle", "Angled side view showing the silhouette",
                    required = true, icon = "footprints",
                ),
                PhotoRole("sole", "Sole", "Outsole / tread — show wear honestly", required = true, icon = "footprints"),
                PhotoRole("tag", "Size Stamp", "Tongue or insole size + brand stamp", required = false, icon = "tag"),
                PhotoRole(
                    "interior", "Insole", "Inside the shoe — footbed condition",
                    required = false, icon = "layers",
                ),
                PhotoRole(
                    "accessory", "Box / Extras", "Original box, spare laces, papers",
                    required = false, icon = "package",
                ),
                PhotoRole(
                    "defect", "Defect", "Tight crop on any flaw — stain, snag, scuff, crack. Be honest.",
                    required = false, icon = "alert-triangle",
                ),
            ),
        )

        val bundledFallback: Map<String, PhotoProfile> = mapOf(
            "clothing" to clothingFallback,
            "shoes" to shoesFallback,
            "other" to genericFallback,
        )
    }
}

/**
 * Fetches + caches the profile table; a single fetch per session (the edge
 * caches server-side too). Bundled profiles guarantee callers can ALWAYS
 * resolve a profile — a fetch failure is non-fatal and retries next session.
 *
 * US-2496: the table is NOT account-blind. `GET /api/flipdesk/photo-profiles`
 * answers per `workspaceOwnerId ?? userId` - a seller who cannot use the
 * authenticity add-on is served fewer roles - so a table fetched for one tenant
 * is the wrong answer for the next one. This store is a process-wide singleton
 * that outlives both a sign-out and a workspace switch, so the table carries
 * the owner it was fetched FOR ([ownerProvider]) and is only ever handed back
 * to that owner. A stale table reads as "not loaded": callers get the bundled
 * fallback and the next [loadIfNeeded] refetches.
 *
 * The stamp is the mechanism deliberately, rather than a reset called from the
 * sign-out and switch paths. A stamp cannot be forgotten by a future exit path,
 * and this cache has three of them already (sign-out, workspace switch, and the
 * involuntary switch when a membership is revoked mid-session).
 */
class PhotoProfileStore(
    private val api: EdgeApi,
    /** Active workspace owner, else the signed-in user; null when signed out. */
    private val ownerProvider: () -> String? = { null },
) {

    @Serializable
    internal data class Wrapper(val profiles: Map<String, PhotoProfile>)

    /** A loaded table plus the tenant it was loaded for. */
    private class Loaded(val owner: String?, val profiles: Map<String, PhotoProfile>)

    @Volatile
    private var loaded: Loaded? = null

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * The server table when it belongs to the CURRENT tenant, else the bundled
     * fallback. Checked on every read rather than only at load time, because
     * the surfaces that display a profile (retag, the photo manager) do not all
     * call [loadIfNeeded] first - a check that only ran on load would leave
     * them showing the previous account's table until capture was opened.
     */
    private val profiles: Map<String, PhotoProfile>
        get() {
            val owner = ownerProvider() ?: return PhotoProfile.bundledFallback
            return loaded?.takeIf { it.owner == owner }?.profiles
                ?: PhotoProfile.bundledFallback
        }

    /** Clothing is the historical default for null/unknown categories. */
    fun profileFor(category: String?): PhotoProfile {
        val table = profiles
        category?.let { table[it] }?.let { return it }
        return table["clothing"] ?: PhotoProfile.clothingFallback
    }

    /**
     * US-2469: resolves the profile for an item, consulting the free-text
     * GARMENT word ("blazer", "dress pants") as well as the item_category.
     *
     * Mirrors `getPhotoProfile(category, garment)` on the edge, including its
     * fallback: when no usable item_category is given, the garment word resolves
     * a group and that group maps back to an item_category profile where one
     * exists. That matters because `items_full` has no item_category column at
     * all, so a free-text word is often the only thing a caller has.
     *
     * The clothing sub-profiles are keyed "clothing:top", "clothing:suit" and
     * friends, so a t-shirt is never offered an inseam slot and a blazer IS
     * offered a shoulder — which `item_category` alone cannot tell apart,
     * because it says "clothing" for both.
     */
    fun profileFor(category: String?, garment: String?): PhotoProfile {
        // One snapshot for the whole resolution: re-reading the tenant between
        // the lookups could mix two tables in one answer.
        val table = profiles
        // An explicit, known item_category wins outright.
        if (!category.isNullOrEmpty() && category != "clothing") {
            table[category]?.let { return it }
        }
        val group = GarmentGroup.from(garment ?: category)
        group.itemCategoryProfileKey?.let { key -> table[key]?.let { return it } }
        table[group.clothingProfileKey]?.let { return it }
        return profileFor(category)
    }

    /**
     * Load once PER TENANT from the edge (1h TTL cache); safe to call
     * repeatedly. Signed out there is no tenant to load for, so this is a no-op
     * and callers keep the bundled fallback.
     */
    suspend fun loadIfNeeded() {
        val owner = ownerProvider() ?: return
        if (loaded?.owner == owner) return
        runCatching {
            val body = api.getRaw("/api/flipdesk/photo-profiles", cacheTtlMillis = 3_600_000)
            val wrapper = json.decodeFromString(Wrapper.serializer(), body)
            // Stamped with the owner resolved BEFORE the request. A switch that
            // landed mid-flight leaves this table stamped for the tenant it was
            // actually fetched for, so the new tenant reads it as stale and
            // refetches - rather than inheriting an answer meant for someone else.
            if (wrapper.profiles.isNotEmpty()) {
                loaded = Loaded(owner, wrapper.profiles)
            }
        } // failure keeps the bundled fallback — retry next session
    }
}
